internal/report/source.go

```diff
@@ -595,18 +602,31 @@ func openSourceFile(path, searchPath string) (*os.File, error) {
 }
 
 // trimPath cleans up a path by removing prefixes that are commonly
-// found on profiles.
-func trimPath(path string) string {
-	basePaths := []string{
-		"/proc/self/cwd/./",
-		"/proc/self/cwd/",
+// found on profiles plus configured prefixes.
+func trimPath(path, trimPath, searchPath string) string {
+	sPath, searchPath := filepath.ToSlash(path), filepath.ToSlash(searchPath)
+	if trimPath == "" && searchPath != "" {
+		// If the trim path is not configured, try to guess it heuristically:
+		// search for basename of each search path in the original path and, if
+		// found, strip everything up to and including the basename. So, for
+		// example, given original path "/some/remote/path/my-project/foo/bar.c"
+		// and search path "/my/local/path/my-project" the heuristic will return
+		// "/my/local/path/my-project/foo/bar.c".
+		for _, dir := range filepath.SplitList(searchPath) {
+			want := "/" + filepath.Base(dir) + "/"
+			if found := strings.Index(sPath, want); found != -1 {
+				return path[found+len(want):]
```

Comment:
I think we want path instead of sPath, as we want to return the system path.

However, maybe we should use FromSlash(sPath). Not sure if there is a guarantee that in all platforms the path separator is a single character.

Reply:
From https://golang.org/src/path/filepath/path.go?s=4426:4458#L155 and from ToSlash documentation that says "ToSlash returns the result of replacing each separator character in path with a slash ('/') character. Multiple separators are replaced by multiple slashes." I am pretty confident that it's safe to slice the original path using the slashized one as the index.

Reply:
I would prefer to see `FromSlash(sPath[found+len(want):])`. Just because the subtle juggling of different path values makes the code harder to read.

Reply:
I think the `trimPath` function could have a policy of converting `path` and `trimPath` using `ToSlash` operating only on these and then explicitly converting _any_ return value using `FromSlash`. This would avoid having to reason about whether the indexes line up between different versions of each path on different OSes.

I agree with @aalexand that the indexes very likely do line up in every case. But I think it is undesirable to have to think so carefully about the correctness of this function. It would be preferable if it was _obviously_ correct.

This has the downside of increasing allocations inside the trimPath function. I profiled the pprof tool using such an explicit `To/FromSlash` policy and found that the allocations in `trimPath` do not appear at all in the profile generated.

For reference my method of profiling is run

`pprof --http :8081 -source_path ${GOPATH}/src:${GOROOT}/src profile_output`

 on an existing server. Then using the web-ui navigate to the `source` view and then take a heap profile of the running `pprof` process. The profile is then viewed with `--sample_index=alloc_objects`.

Reply:
I too would prefer to use FromSlash on the return, if only for symmetry.

Reply:
IMO the documentation for ToSlash and FromSlash clearly says that these functions only replace character for character, so I think there is nothing to worry about. Function name is `trimPath` so I think it's only job is trimming. I can rename it to trimAndFromSlashPath and make the suggested change if it's everybody's preference (not mine).

Reply:
Note that trimPath call is also used to trim the paths before displaying in the graph or source / assembly view, so ToSlash'ing the path here would make Unix paths display with backslash if someone opens on a Windows machine a profile collected on Linux. This is an edge case but I think the behavior as coded now would be more reasonable.

Reply:
That's a good point. Since running `FromSlash(ToSlash(somePath))` won't always produce a string identical to `somePath` then I am happy with the function as it is.

Reply:
SGTM. Thanks